### ISSUE TITLE
Should Semaphore come in a 'bounded' flavor?

### ISSUE THREAD

**Issue by kalexmills** [typelevel/cats-effect#823]
As currently implemented, `Semaphore` represents an unbounded pool of permits, which allows for the below behavior, when trying to use it as a lock.

```
for {
  lock <- Semaphore[IO](1)
  _ <- lock.releaseN(1)  // now there are two permits in the pool...
  _ <- lock.acquireN(1) // does not block
  _ <- lock.acquireN(1) // does not block either,
```

For now, the only way to safely use a `Semaphore` is to ensure that each `release` is preceeded by an `acquire`. This fits most use-cases, but after glancing at the implementation, I think it ought to be possible to enable bounded semaphores by [generalizing the addition used in `releaseN`](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala#L256).

```
                case Right(m) => Right(m + n)
```

I'm not sure if this should be generalized, or what the right level of generalization would be, but I'm happy to imagine an option.

If we let clients replace this `m + n` with a custom `f: (Int, Int) => Int`, I'd hazard a guess that the only danger here is not unblocking the queue in the case that `m = 0` and `n > 0`. For a bounded Semaphore we would replace this with `(m,n) => Math.min(m + n, maxPermits)`.

... I had a use-case for this with hand-over-hand locking, but I've decided it admits race-conditions so I'll need to use `Deferred` as a reentrant lock instead. I still figured I'd raise the topic in case there are other use-cases for a bounded Semaphore.

**Comment by SystemFw** [comment]
I think the rationale is to closely follow the concept of Semaphore as per the concurrency literature(`release` is `P` and `increment` is `V`). More complex behaviour can be obtained by composing semaphores (tons of literature on that), although I think `Ref + Deferred` forms a nicer basis for complex structure since blocking abstractions like `Semaphore` are tricky.

For the simple cases where you are just using `Semaphore` as a lock, you can use `.withPermit` instead, although `Semaphore`s overall are more general than locks

**Comment by SystemFw** [comment]
Btw I didn't necessarily mean that the issue should be closed :)
Just documenting where the current design comes from

**Comment by kalexmills** [comment]
Thanks for clarifying. 

The use-case I was interested in can be seen [in this gist](https://gist.github.com/kalexmills/856181d7f9f090aa5e11952cbf20c895). I've since learned that the proposal this issue represents will not resolve the issue I was having. I was doing something incorrectly and actually did not need to unlock twice.

In any case, `MVar::tryPut` would work, but a binary semaphore might(?) be lighter-weight, depending on the internals.

There may be other use-cases which require something like this so I guess we can let others weigh in.
